gui-lib/framework/private/text-port.rkt

```diff
@@ -538,50 +544,159 @@
+                 ((empty-markup? markup) (void))
+                 ((horizontal-markup? markup)
+                  (for-each (lambda (markup)
+                              (insert-markup markup text style #t))
+                            (horizontal-markup-markups markup)))
+                 ((vertical-markup? markup)
+                  (if inline?
+                      (send text insert (markup->snip markup style #f))
+                      (for-each/between (lambda (markup)
+                                          (insert-markup markup text style #f))
+                                        (lambda () (send text insert #\newline))
+                                        (vertical-markup-markups markup))))
+                 ((srcloc-markup? markup)
+                  (insert-srcloc-markup markup text style))
+                 ((framed-markup? markup)
+                  (send text insert (markup->snip (framed-markup-markup markup) style #t)))
+                 ((image-markup? markup)
+                  (send text insert (image-markup->snip markup style)))))
+             
+             (define (image-markup->snip markup style)
+               (let ((data (image-markup-data markup)))
+                 (cond
+                   ((is-a? data snip%) data)
+                   ((snip-special? data)
+                    (define snip (snip-special->snip data))
+                    (if (exn:fail? snip)
+                        (markup->snip (image-markup-alt-markup markup) style #f)
+                        snip))
+                   (else
+                    (markup->snip (image-markup-alt-markup markup) style #f)))))
+             
+             (define (record-dc-datum->bitmap datum width height)
+               (with-handlers ((exn? (lambda (e) #f)))
+                 (let ((proc (recorded-datum->procedure datum))
+                       (bitmap (make-object bitmap% width height)))
+                   (let ((dc (new bitmap-dc% [bitmap bitmap] )))
+                     (proc dc)
+                     bitmap))))
+        
+             (define (insert-srcloc-markup srcloc-markup text style)
+               (let ((start (send text get-end-position)))
+                 (insert-markup (srcloc-markup-markup srcloc-markup) text style #t)
+                 (let ([end (send text get-end-position)])
+                   (send text set-clickback
+                         start end
+                         (lambda (t s e)
+                           (srcloc-snip:select-srcloc (srcloc-markup-srcloc srcloc-markup))))
+                   (send text change-style
+                         (make-object style-delta% 'change-underline #t)
+                         start end #f))))
+             
+             ; like for-each, but with a thunk that gets called between elements
+             (define (for-each/between proc between list)
+               (let loop ((list list))
+                 (cond
+                   ((null? list) (values))
```

Comment:
Maybe `(void)` instead of `(values)`?

Reply:
Done. (Even though I really like `values` better here, so that no one gets tempted to do anything with the return value.)

Reply:
Thanks. `(void)` seems to be the convention elsewhere is all so it might confuse. I wonder which would work better for useful error messages when a bug happens.